### ISSUE TITLE
Compilation error message readability

### ISSUE THREAD

**Issue by dhollman** [kokkos/kokkos#2088]
Kokkos uses a *lot* of SFINAE on detection of interfaces, particularly with things like `parallel_reduce`.  Besides being completely unreadable at the source code level (I've been working on Kokkos for a while now, and between the wiki and the source code, I still can't discern the full set of allowed overloads for `parallel_reduce`), the overload set of `parallel_reduce` could also be re-written to produce **much** friendlier error messages (especially in modern compilers).  I'm working through rewriting some of it to fix #2034 anyway, so I thought I'd do a bit more modernization while I'm at it.  I've worked out the best way to produce readable error messages for SFINAE overloads in modern GCC and clang (it appears everything else is pretty much a lost cause), but the problem is that it's different for the two compilers.  I've prepared a [godbolt demo](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAKxAEZSAbAQwDtRkBSAJgCFufSAZ1QBXYskwgA5BwAMAQW4BmPC2QMRWANQclAYQIBPAA6YA%2BgWJNCg3djmKFAeidaAygHdCyBFoBGmAQemJgsWrRarOhaXFpEWoIhOs6uaAC2xnjMBHioYaIExiIEWqgAZlr4ZWWYxKEEDi6RxsbEqEw%2BmLYKylhlqphaegDyALIACgCSADLYZgDqABKTeotm8uPjAErD8qvhDo1OugAip%2BdKZ5cXV7c399ePd08Pz2%2BvHy9f79cpWsPGHJ5cIgLRpEiDOpMdBMPwMQaqMEYWphADiej0hy4KgqIwmMzmSxWaw2212%2BxuBx6CkElhEyBKABUuiUOAB2HhaABuqDw0TKqFQEFUBAAlLpOeyrnxqfICJgMtlBro9OomIJBFpGXYEnSGWCmABrcx1ACOIjwdTS9S0oNp6BAIDp5iMph0HPd0sOfy2mHNloV9U1LFQJWtJWQUT5THlDnlipjyv0ao1WrsDiYxVQWjKIjUEEZcVIWhF4oUWi0IcwAA8JICIFXa5h6wQAHQCoWirsOCsAWh1rtCTGtuodICHcPMeDKKp7Ffno8deEEZjQxkMq7ytOI9Jyk5V2qU2EdXKYGkwc4XOi4ADZuDeDcazGaLVb6iqsOpBxA2x2IN2jxPM8REwUhLwXEVwJ0QCnRMC8eg9OoCDEMJaAlT1vXkJp5jaNhKxENIAmIUoKiYYhgAIoMtEjFh8BhWMFHjYwlWg1VmFTQ97AUVQGAGEsWBKXN80ZMtFEQwIUJidCpUwkU4hZf93RleQKyQyShOQAsWXZPg2ROYsb3FJRlJk2UXE%2BH5LIs6zvhsqzbIc%2Byzj%2BAEgTCLhQXBOotChGFJ34pEsGIMI1TYLElEwXjcTGKZZgWZZVnWTYdj2RZoNuKTZW3XctRZJTuV5flBWFASjMlPSJUwpiWJVFNNU4rQREEVRgC0BB1TMFgRAYZhiA3DszHDBAMDMMjOkHdLKkwL84Ige1HU/BhTwYA87H/dtioAkzGIVZjE1Y6r9tq9jBFbM6dTqrQADFi0u%2BQGA8JhDEEAA1QrbpOs7Wy0eRyNsI8HGy/VfNheEzGXMwsHlBlMGiO0CDHMoz0SCw4PymTjKq3aav0Q75VYuqvouk7ro%2B9VTrOtMAZpPUShBydwZXKHpvldAVRuxcQB5PkLHZg8iYBo9i0ZAXsFtTnnVRt0dIwzHZTxpM2PJqmuPkIGSnalcup6sj%2BsFQbAmGuGfMwaFQanJnAhZ2GVU1zrut6vXUANggjdG8QEECODi259BhfTBDyq9MzXF9f03wE4NQzBQJqKjej4LlbGjuTEnOIzLMczzTTCwIYtSznRs6wIBtUBrYufw2ztu3LLR%2ByPOI4JYYdBnm8dm4Z6dZ1rq824htcNzQFh1bwfd9E4oDzygit73vNqOu1x2zAGoaMFWmDlpAsCe4ggSoLsR1B0OcTkOCg5jIwkOtBwvJWq6wjahIyJyMoyO49o6MGKThN8eO5X0%2B4iwXiLAEQCSzsJUSMs1Jny4NJCqso5LylpIpHSc5oFhA0lpWkqC9IGTKsfZyjgsLHEcnZMhpCKFOXIVQ2yPRsShCqA4KQopGDSAAKxSFICwaQshOGoGkJiXg/AEiiHEMqbEtBOEEB4cwlhhoQBKAAJytgACwAA4uBqLZIo2gsg2FsK4LIG8KjWFSBUZw7hUheGkH4VIThggQCyFINIqxzDSBwFgDARAKBUAZCyLUcglB0iZHhMQFAzA2BqNkE4/oDB5TEAcRAPwMjSB%2BFULraQkjSDpGtAJYYQDDApKwGkVgwB4RFIDAyPAXIugpJrNNYokgpBZJFJFFJvE/BWD6noZEmSpHEDwGkPpLCInsH4PwRgo8HGQBYagQEuRh7SF7MMJQdd7SnE4EI3gtA2T2NERIOgIz2EWJSbY6saiby9mMdRUZWg1GtlkA8rQEBcCEBINeJQ9Ahi%2BJCY/ZQtBRRDHGbwKRMiuykHkUoFRrYlBslvGyNhnyVE3jhWwxRpjzGkCGZo1sKLZBKBvLIWguiolKCUNiLhpzpD2Mcc4sF7ivEQCQAUIoecKAQGCf4sJ9BYZvO5aQDwVhjD0oYMcylribHSH%2BVoLwrstDnMudc0KrV7mPOsS43h4LPbQlqJQI5ZjOFDLZLQGFXBFE3iUWotROjzUovFdY2xNKnEatkRCkAXA2StloGo2FWi2E3mJcioxpilAnIlY6ulriWEeKgIypAnLQmBI5T8rlIBgA%2BtILE%2BJiTkkSrSc3PqfTsm%2BNyQQfJDBCkSuKaU8pVbKk5BqQ4iV9TkCNKLa00VEqOldMMD0rARbLCDOGYwUpKBgUCA6dM/8Nj5lbiWSstZCMNnjp2XssQByAWmI4favh0gFVXJUVoYAyBkB3Jhc815RBiL/OLD0vxoSPlcEBYIvgILI2arkQopQMKyW/r/X%2BjFhqFGPKtcY28aKPWKLZGyDNliHXUqELSl10a40gBZcUJNCbdVcFILyq9dABVCuHZ27dcHd0yGxLEWVvh93XOPae%2B5ob32uu1UFPVgGsUgBvMo4lSg2FsihbIOFsguBcHoGRyVdjEPOrBZ%2BslP7/2KaUCGsN8GpPIf1VwIDuinESYjRp0gNSEkLJACooAA%3D%3D%3D) illustrating the problem.  Basically, gcc produces better error messages for something of the form of a "direct" substitution failure (see demo), whereas clang produces better error messages for something that looks like a concept requirement.

So the question is this: if we're going to rewrite some of this stuff, which approach should we favor?  In the long run, the concept requirements style will have the most support, since compilers' implementations of concepts will be moving into production soon.  However, my understanding is that Kokkos has more gcc users right now (and for the foreseeable future), and I think the gcc elucidation of a "direct" substitution failure is much more readable in the short term.  Thoughts?

**Comment by crtrott** [comment]
I am somewhat against the two alternatives methods, it will add to the testing burding significantly ...

**Comment by mhoemmen** [comment]
Which approach would tend to make error messages shorter?  Users get overwhelmed mainly by the length of error output, not necessarily by its crypticness.  Shorter error messages also reduce burden on test machines (that's mainly an issue with warnings, though -- in some cases we have GB of output).

**Comment by dhollman** [comment]
@crtrott Can you clarify? Either version should decrease compilation time significantly compared to the current `Impl::FunctorAnalysis` approach.

The kinds of compilation bugs I had in mind are things like:

```c++
struct MyReducer {
  using reduecr = MyReducer; // oops, typo
  /* ... */
};

int x;
Kokkos::parallel_reduce(42, Functor{}, MyReducer{x});
```

the error for which is nearly impossible to understand in the current implementation.  Even simple things like:

```c++
const int x = 0; // oops
Kokkos::parallel_reduce(42, Foo{}, x);
```

don't give a missing overload of `parallel_reduce`. (Actually, the latter example just compiles and runs and gives nonsense, which we **really** need to fix. I'll open up another bug.)

@mhoemmen The clang approach (approach 2 in the demo) gives shorter error messages in both cases, but there may be information missing that would make these above bugs difficult to track down (like the fact that the misspelling of `reducer` is why the `Reducer` overload was not found; all that it would show is something like `constraint not met: is_reducer_type<Reducer>::value`).  Also, if error message length is the issue, might I suggest that we make `-fmax-errors=1` a default (or at least a suggested) compiler argument?

**Comment by crtrott** [comment]
What I meant is if we have two variants of doing stuff (one with error check one wihtout) we need to test both paths too.

**Comment by dhollman** [comment]
@crtrott I agree with that.  But then how do we decide where to fall on the "error readability" versus "compilation cost"? Within reason, which one takes priority?

**Comment by crtrott** [comment]
We just expect that everyone write bug free code ;-)

Seriously though: I don't have a great answer unfortunately. We need to see what the impact is.

**Comment by ajpowelsnl** [comment]
@dalg24 -- can we close this issue, or should it be re-assigned?
